File: solutions/python/matching-brackets/1/matching_brackets.py
pairs_dict = {
    "(": ")",
    "{": "}",
    "[": "]"
}


def is_paired(input_string):
    brackets_stack = []
    for char in input_string:
        if char in "([{":
            brackets_stack.append(char)
        elif char in ")]}":
            if len(brackets_stack) == 0:
                return False

            last = brackets_stack.pop()
            if char != pairs_dict[last]:
                return False

    return False if len(brackets_stack) else True

File: solutions/typescript/matching-brackets/1/matching-brackets.ts
type OpeningBracket = "(" | "{" | "[";
type ClosingBracket = ")" | "}" | "]";

const bracketPairs: { [K in OpeningBracket]: ClosingBracket } = {
  "(": ")",
  "{": "}",
  "[": "]",
};

const isOpening = (bracket: string): boolean => "({[".includes(bracket);
const isCLosing = (bracket: string): boolean => ")}]".includes(bracket);

export function isPaired(input: string): boolean {
  const bracketStack: OpeningBracket[] = [];

  for (const c of input.split("")) {
    if (isOpening(c)) {
      bracketStack.push(c as OpeningBracket);
    } else if (isCLosing(c)) {
      const last: OpeningBracket | undefined = bracketStack.pop();

      if (!last || c !== bracketPairs[last]) return false;
    }
  }

  return bracketStack.length === 0;
}
